econml/iv/dml/_dml.py

```diff
@@ -15,78 +15,280 @@
+
+        else:
+            if hasattr(self._model_z_xw, 'score'):
+                Z_X_score = self._model_z_xw.score(X=X, W=W, Target=Z, sample_weight=sample_weight)
+            else:
+                Z_X_score = None
+            return Y_X_score, T_X_score, Z_X_score
+
+    def predict(self, Y, T, X=None, W=None, Z=None, sample_weight=None, group=None):
+        Y_pred = self._model_y_xw.predict(X=X, W=W)
+        T_pred = self._model_t_xw.predict(X=X, W=W)
+
+        if self._projection:
+            # concat W and Z
+            WZ = self._combine(W, Z, Y.shape[0])
+            T_proj = self._model_t_xwz.predict(X, WZ)
+        else:
+            Z_pred = self._model_z_xw.predict(X=X, W=W)
 
-class _BaseDMLATEIVModelFinal:
-    def __init__(self):
-        self._first_stage = LinearRegression(fit_intercept=False)
-        self._model_final = _FinalWrapper(LinearRegression(fit_intercept=False),
-                                          fit_cate_intercept=True, featurizer=None, use_weight_trick=False)
+        if (X is None) and (W is None):  # In this case predict above returns a single row
+            Y_pred = np.tile(Y_pred.reshape(1, -1), (Y.shape[0], 1))
+            T_pred = np.tile(T_pred.reshape(1, -1), (T.shape[0], 1))
+            if not self._projection:
+                Z_pred = np.tile(Z_pred.reshape(1, -1), (Z.shape[0], 1))
+
+        Y_res = Y - Y_pred.reshape(Y.shape)
+        T_res = T - T_pred.reshape(T.shape)
+
+        if self._projection:
+            Z_res = T_proj.reshape(T.shape) - T_pred.reshape(T.shape)
+        else:
+            Z_res = Z - Z_pred.reshape(Z.shape)
+        return Y_res, T_res, Z_res
+
+
+class _OrthoIVModelFinal:
+    def __init__(self, model_final, featurizer, fit_cate_intercept):
+        self._model_final = clone(model_final, safe=False)
+        self._original_featurizer = clone(featurizer, safe=False)
+        self._fit_cate_intercept = fit_cate_intercept
+
+        if self._fit_cate_intercept:
+            add_intercept_trans = FunctionTransformer(add_intercept,
+                                                      validate=True)
+            if featurizer:
+                self._featurizer = Pipeline([('featurize', self._original_featurizer),
+                                             ('add_intercept', add_intercept_trans)])
+            else:
+                self._featurizer = add_intercept_trans
+        else:
+            self._featurizer = self._original_featurizer
+
+    def _combine(self, X, T, fitting=True):
+        if X is not None:
+            if self._featurizer is not None:
+                F = self._featurizer.fit_transform(X) if fitting else self._featurizer.transform(X)
+            else:
+                F = X
+        else:
+            if not self._fit_cate_intercept:
+                raise AttributeError("Cannot have X=None and also not allow for a CATE intercept!")
+            F = np.ones((T.shape[0], 1))
+        return cross_product(F, T)
```

Comment:
Yeah I think this is hard since this final wrapper is in the same layer of _rlearner final model wrapper, not the dml one, so the fit function doesn't have the same signature. If only create a base class to have `_init_()` and `_combine()`, where should I put it? since only these two places have this similar implementation, I will leave it for now unless you have any better ideas?